Important updates in regard to Microservice architecture patterns chapter and s…

CESARDELATORRE opened the pull request:
# Important updates in regard to Microservice architecture patterns chapter and some updates about products.

## Summary

There was a section that was a different topic included as part of another topic's page, so I split it.
For that I needed to re-number the images and while doing so, I was reviewing all that content and updating things that needed to be updated either because of changes in products or more accurate technical explanations.

Fixes #Issue_Number
This is not related to an specific issue. I'm starting to update this documentation so it is up-to-date plus in the upcoming days I will update code snippets so it also supports .NET Core 2.0 (current content shows examples in .NET 1.1)

## Suggested Reviewers

@BillWagner

CESARDELATORRE committed:
Important updates in regards Microservice architecture patterns and some updates about products.

CESARDELATORRE committed:
Important updates in regards Microservice architecture patterns and some updates about products.

CESARDELATORRE committed:
Important updates in regards Microservice architecture patterns and some updates about products.

dnfclas commented:
@__CESARDELATORRE__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all .NET Foundation-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

mairaw commented:
One quick note (I haven't reviewed the whole thing yet): I see that you're renaming a file. When we do that, we need to add an entry here to redirect traffic from the old page to the new one here:
https://github.com/dotnet/docs/blob/master/.openpublishing.redirection.json

In this case, you need to set the three properties.

CESARDELATORRE commented:
OK, I just did here:

https://github.com/dotnet/docs/pull/3408


It is a different PR. Is that right?


Thanks,

Cesar.



________________________________
From: Maira Wenzel <notifications@github.com>
Sent: Friday, October 13, 2017 5:45:01 PM
To: dotnet/docs
Cc: Cesar De la Torre Llorente; Author
Subject: Re: [dotnet/docs] Important updates in regard to Microservice architecture patterns chapter and s… (#3407)


One quick note (I haven't reviewed the whole thing yet): I see that you're renaming a file. When we do that, we need to add an entry here to redirect traffic from the old page to the new one here:
https://github.com/dotnet/docs/blob/master/.openpublishing.redirection.json<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fdocs%2Fblob%2Fmaster%2F.openpublishing.redirection.json&data=02%7C01%7Ccesardl%40microsoft.com%7C93669538b4aa4d371f8808d5129ccf65%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636435387058603725&sdata=CI8hl67tZkofB0ksKZiT%2BJ%2BllLwiqB59ki3wRRS65VQ%3D&reserved=0>

In this case, you need to set the three properties.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fdocs%2Fpull%2F3407%23issuecomment-336597145&data=02%7C01%7Ccesardl%40microsoft.com%7C93669538b4aa4d371f8808d5129ccf65%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636435387058603725&sdata=voYW4JnearKCFMRIh%2BPPERxcFW1iFD0%2Be2qnZSGLCaI%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABoh-xVPLuD3H_fBLVhiuBx5U5fsiSBlks5ssAQNgaJpZM4P5Mc8&data=02%7C01%7Ccesardl%40microsoft.com%7C93669538b4aa4d371f8808d5129ccf65%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636435387058603725&sdata=%2BiqjODu%2BzujsrtgomZGL16Zlv1xqi7ZRfrKgKRGJrFQ%3D&reserved=0>.

mairaw commented:
It needs to be in the same otherwise it will complain the file still exists. Let me add it here and I'll close the other PR.

mairaw committed:
added redirect

mairaw committed:
added redirect

mairaw committed:
added redirect

mairaw commented:
@CESARDELATORRE done. Also, make sure you follow the process to sign the CLA.

CESARDELATORRE committed:
Merge remote-tracking branch 'upstream/master'

CESARDELATORRE committed:
Revert so PR is not mixing multiple eBooks

CESARDELATORRE committed:
Revert so PR is not mixing multiple eBooks

CESARDELATORRE committed:
Merge branch 'master' of https://github.com/dotnet-architecture/docs

CESARDELATORRE committed:
Update orchestrate-high-scalability-availability.md

CESARDELATORRE commented:
I removed the minor change in the other "eBook" content (Docker lifecycle ebook) so we don't mix different content in the same PR.
It was a minor change that I did without noticing, although the change was right (Document DB --> Cosmos DB), but I prefer not to mix the PR across multiple "ebooks".

dnfclas commented:
__@CESARDELATORRE__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

mairaw committed:
add 4.7.1 to API Browser quick filter (#3445)

mairaw committed:
minor fixes (#3434)

mairaw committed:
Add new ASP.NET 4.7.1 APIs (#3439)

* added new ASP.NET 4.7.1 APIs

* removed extra spaces

mairaw committed:
replace assetids with xref links (#3443)

CESARDELATORRE committed:
Updates on .NET Core 2.0 and news in Docker in chapters related to Docker and framework selection in .NET

CESARDELATORRE committed:
Merge branch 'master' of https://github.com/dotnet-architecture/docs

xaviex committed:
Fixing FWLinks in Core and FSharp Docs folders (#3441)

* Core/FSharp FWLink Finishes

* Revert "Core/FSharp FWLink Finishes"

This reverts commit b2ebe95289dfbc9e54e48e5574bf1f73dd102923.

* Fixed FWLinks to include final redirects

* Fixed Visible Link

* Fixed Localizing

* feedback

CESARDELATORRE committed:
Updates in State and Data in Docker apps

CESARDELATORRE committed:
Updates in State and Data in Docker apps

CESARDELATORRE committed:
Minor changes in Architecture chapter/section

CESARDELATORRE committed:
Minor changes in Architecture chapter/section

CESARDELATORRE committed:
Updates for the Docker development process chapter

CESARDELATORRE committed:
Updates for the Docker development process chapter

mairaw commented on docs/standard/microservices-architecture/architect-microservice-container-applications/communication-in-microservice-architecture.md:
```diff
@@ -1,15 +1,15 @@
-# Communication between microservices
+# Communication in a microservice architecture
 
 In a monolithic application running on a single process, components invoke one another using language-level method or function calls. These can be strongly coupled if you are creating objects with code (for example, new ClassName()), or can be invoked in a decoupled way if you are using Dependency Injection by referencing abstractions rather than concrete object instances. Either way, the objects ...
```
inline code `new ClassName())`

CESARDELATORRE replied:
OK, I'm updating that.

mairaw commented on docs/standard/microservices-architecture/architect-microservice-container-applications/communication-in-microservice-architecture.md:
```diff
@@ -41,13 +41,17 @@ A microservice-based application will often use a combination of these communica
 If possible, never depend on synchronous communication (request/response) between multiple microservices, not even for queries. The goal of each microservice is to be autonomous and available to the client consumer, even if the other services that are part of the end-to-end application are down or unhealthy. If you think you need to make a call from one microservice to other microservices (like p ...
 
-Moreover, having dependencies between microservices (like performing HTTP requests between them for querying data) not only makes your microservices not autonomous. In addition, their performance will be impacted. The more you add synchronous dependencies (like query requests) between microservices, the worse the overall response time will get for the client apps.
+Moreover, having HTTP dependencies between microservices (like creating long request/response cycles with HTTP request chains as shown in the first part of the Figure 4-15), not only makes your microservices not autonomous. In addition, their performance will be impacted. The more you add synchronous dependencies (like query requests) between microservices, the worse the overall response time wil ...
```
not only makes your microservices not autonomous _> this sentence seems to be cut
I'd expect a construction such as `not only ... but also`

suggestion: not only makes your microservices not autonomous but also their performance is impacted.

Avoid having sentences in parentheses and future tense.

suggestion:
The more you add synchronous dependencies between microservices, such as query requests, the worse the overall response time gets for the client apps.

CESARDELATORRE replied:
OK, updated according your suggestions.

mairaw commented on docs/standard/microservices-architecture/architect-microservice-container-applications/docker-application-state-data.md:
```diff
@@ -23,10 +23,10 @@ The following solutions are used to manage persistent data in Docker application
-
 -   [Azure Storage](https://docs.microsoft.com/azure/storage/), which provides geo-distributable storage, providing a good long-term persistence solution for containers.
 
+-   Remote relational databases like [Azure SQL Database](https://azure.microsoft.com/en-us/services/sql-database/) or NoSQL databases like [Azure Cosmos DB](https://docs.microsoft.com/en-us/azure/cosmos-db/introduction), or cache services like [Redis](https://redis.io/).
```
remove locale from URLs

CESARDELATORRE replied:
OK, fixed.

mairaw commented on docs/standard/microservices-architecture/architect-microservice-container-applications/index.md:
```diff
@@ -15,15 +15,15 @@ ms.topic: article
 When you design a container image, you will see an [ENTRYPOINT](https://docs.docker.com/engine/reference/builder/) definition in the Dockerfile. This defines the process whose lifetime controls the lifetime of the container. When the process completes, the container lifecycle ends. Containers might represent long-running processes like web servers, but can also represent short-lived processes lik ...
 
-If the process fails, the container ends, and the orchestrator takes over. If the orchestrator was configured to keep five instances running and one fails, the orchestrator will create another container instance to replace the failed process. In a batch job, the process is started with parameters. When the process completes, the work is complete.
+If the process fails, the container ends, and the orchestrator takes over. If the orchestrator was configured to keep five instances running and one fails, the orchestrator will create another container instance to replace the failed process. In a batch job, the process is started with parameters. When the process completes, the work is complete. This guidance drills-down on orchestratros, later  ...
```
typo: orchestratros

CESARDELATORRE replied:
typo fixed.

mairaw commented on docs/standard/microservices-architecture/architect-microservice-container-applications/scalable-available-multi-container-microservice-applications.md:
```diff
@@ -13,31 +13,41 @@ ms.topic: article
 From an architecture and development point of view, if you are building large enterprise composed of microservices-based applications, it is important to understand the following platforms and products that support advanced scenarios:
 
-**Clusters and orchestrators**. When you need to scale out applications across many Docker hosts, as when a large microservice-based application, it is critical to be able to manage all those hosts as a single cluster by abstracting the complexity of the underlying platform. That is what the container clusters and orchestrators provide. Examples of orchestrators are Docker Swarm, Mesosphere DC/OS ...
+**Clusters and orchestrators**. When you need to scale out applications across many Docker hosts, as when a large microservice-based application, it is critical to be able to manage all those hosts as a single cluster by abstracting the complexity of the underlying platform. That is what the container clusters and orchestrators provide. Examples of orchestrators are Azure Service Fabric, Kubernet ...
```
,  (the last three available through Azure Container Service). -> These orchestrators are available through Azure Container Service.

CESARDELATORRE replied:
But, Service Fabric is NOT available through ACS... I'm changing the way, though.

mairaw replied:
Sorry, I misread and only saw 3 in the list.

mairaw commented on docs/standard/microservices-architecture/architect-microservice-container-applications/using-azure-service-fabric.md:
```diff
@@ -23,67 +23,68 @@ Service Fabric provides two broad areas to help you build applications that use
+**Figure 4-29**. Business microservice mapped to a Service Fabric application with containers and stateful services
 
-Support is also different depending on whether you are using Docker containers on Linux or Windows Containers. Support for containers in Service Fabric will be expanding in upcoming releases. For up-to-date news about container support in Azure Service Fabric, see [Service Fabric and containers](https://docs.microsoft.com/azure/service-fabric/service-fabric-containers-overview) on the Azure websi ...
+For up-to-date news about container support in Azure Service Fabric, see [Service Fabric and containers](https://docs.microsoft.com/azure/service-fabric/service-fabric-containers-overview) on the Azure website.
```
It's actually the docs website. Do you need to say where?

CESARDELATORRE replied:
OK, updated.

mairaw commented on docs/standard/microservices-architecture/docker-application-development-process/docker-app-development-workflow.md:
```diff
@@ -125,24 +125,36 @@ You can specify additional configuration settings in the Dockerfile, depending o
+### Using multi-arch image repositories
 
-A single repo can contain platform variants, such as a Linux image and a Windows image. This feature allows vendors like Microsoft (base image creators) to create a single repo to cover multiple platforms. For example, the [microsoft/dotnet](https://hub.docker.com/r/microsoft/aspnetcore/) repository available in the Docker Hub registry provides support for Linux and Windows Nano Server by using t ...
+A single repo can contain platform variants, such as a Linux image and a Windows image. This feature allows vendors like Microsoft (base image creators) to create a single repo to cover multiple platforms (i.e. Linux and Windows). For example, the [microsoft/dotnet](https://hub.docker.com/r/microsoft/aspnetcore/) repository available in the Docker Hub registry provides support for Linux and Windo ...
```
i.e. -> that is

CESARDELATORRE replied:
ok, updated.

mairaw commented on docs/standard/microservices-architecture/docker-application-development-process/index.md:
```diff
@@ -19,16 +19,18 @@ ms.topic: article
-**Visual Studio with Tools for Docker**. If you are using Visual Studio 2015, you can install the [Visual Studio Tools for Docker](https://marketplace.visualstudio.com/items?itemName=MicrosoftCloudExplorer.VisualStudioToolsforDocker-Preview) add-in. If you are using Visual Studio 2017, tools for Docker are already built-in. In either case, the tools for Docker let you develop, run, and validate y ...
+**Visual Studio (for Windows)**. When developing Docker-based applications with Visual Studio, it is recommended to use Visual Studio 2017 or higher which comes with tools for Docker are already built-in. The tools for Docker let you develop, run, and validate your applications directly in the target Docker environment. You can press F5 to run and debug your application (single container or multi ...
+
+**Visual Studio for Mac**. It is an IDE, evolution of Xamarin Studio, running in macOS and supports Docker since mid-2017. This should be the preferred choice for developers working in Mac machines who also want to use a powerful IDE.
```
is the mid-2017 info relevant?

CESARDELATORRE replied:
OK, changed.

mairaw commented on docs/standard/microservices-architecture/docker-application-development-process/index.md:
```diff
@@ -19,16 +19,18 @@ ms.topic: article
 Whether you prefer a full and powerful IDE or a lightweight and agile editor, Microsoft has tools that you can use for developing Docker applications.
 
-**Visual Studio with Tools for Docker**. If you are using Visual Studio 2015, you can install the [Visual Studio Tools for Docker](https://marketplace.visualstudio.com/items?itemName=MicrosoftCloudExplorer.VisualStudioToolsforDocker-Preview) add-in. If you are using Visual Studio 2017, tools for Docker are already built-in. In either case, the tools for Docker let you develop, run, and validate y ...
+**Visual Studio (for Windows)**. When developing Docker-based applications with Visual Studio, it is recommended to use Visual Studio 2017 or higher which comes with tools for Docker are already built-in. The tools for Docker let you develop, run, and validate your applications directly in the target Docker environment. You can press F5 to run and debug your application (single container or multi ...
```
When developing Docker-based applications with Visual Studio, it is recommended to use Visual Studio 2017 or higher which comes with tools for Docker are already built-in. ->
To develop Docker-based applications, use Visual Studio 2017 or later versions that comes with tools for Docker already built-in.

Docker based apps -> Docker-based apps

CESARDELATORRE replied:
Fixed.

mairaw commented on docs/standard/microservices-architecture/net-core-net-framework-containers/container-framework-choice-factors.md:
```diff
@@ -43,17 +43,17 @@ There are several features of your application that affect your decision. You sh
-    - Your .NET implementation choice is *.NET Framework*, or *.NET Core (future release)*.
-    - Your container platform choice must be *Windows containers* because of the .NET Framework dependency.
-    - When **SignalR services** run on *.NET Core*, you can also choose *Linux containers*.
+    - Your .NET implementation choice is *.NET Framework*, or *.NET Core 2.1 or beyond*.
```
or beyond -> or later versions 

should you also point out that this version is not available yet?

CESARDELATORRE replied:
ok, fixed.

mairaw commented on docs/standard/microservices-architecture/net-core-net-framework-containers/container-framework-choice-factors.md:
```diff
@@ -43,17 +43,17 @@ There are several features of your application that affect your decision. You sh
-    - When **SignalR services** run on *.NET Core*, you can also choose *Linux containers*.
+    - Your .NET implementation choice is *.NET Framework*, or *.NET Core 2.1 or beyond*.
+    - Your container platform choice must be *Windows containers* if you chose the .NET Framework dependency.
+    - When **SignalR services** run on *.NET Core*, you can use *Linux containers or Windows Containers*.
```
expectation of when this will happen?

CESARDELATORRE replied:
Not sure if this date is public or even confirmed, yet, so I prefer not to say it.

mairaw replied:
It seems to be public. See https://github.com/aspnet/Home/wiki/Roadmap#210-roadmap and there's even an alpha release already: https://blogs.msdn.microsoft.com/webdev/2017/09/14/announcing-signalr-for-asp-net-core-2-0/

CESARDELATORRE replied:
Yeah, I know that SignalR ALPHA is public. What is not confirmed is when .NET Core 2.1 will be RTM. In the roadmap it says "Q1-Q2 2018"... Not sure if we want to put such a broad timeframe in the doc. I think we should just say it'll come in .NET Core 2.1. :)

mairaw commented on docs/standard/microservices-architecture/net-core-net-framework-containers/container-framework-choice-factors.md:
```diff
@@ -43,17 +43,17 @@ There are several features of your application that affect your decision. You sh
     - Your .NET implementation choice is *.NET Framework*, or *.NET Core (in the roadmap for a future release)*.
     - Your container platform choice must be *Windows containers* because of the .NET Framework dependency.
-    - When the dependency runs on *.NET Core*, you can also choose *Linux containers*.
+    - If WCF supports *.NET Core* in the future, you could choose between *Linux containers and Windows containers*.
```
but this is not true now except for WCF client no? So why say that?

CESARDELATORRE replied:
Agree. I removed that point.

mairaw commented on docs/standard/microservices-architecture/net-core-net-framework-containers/general-guidance.md:
```diff
@@ -25,15 +25,23 @@ In short, when you create new containerized .NET applications, you should consid
 -   You need to use third-party .NET libraries or NuGet packages that are not available for .NET Core.
 
-Using .NET Framework on Docker can improve your deployment experiences by minimizing deployment issues. This "lift and shift" scenario is important for "dockerizing" legacy applications (at least, those that are not based on microservices).
+Using .NET Framework on Docker can improve your deployment experiences by minimizing deployment issues. This [*"lift and shift" scenario*](https://aka.ms/liftandshiftwithcontainersebook) (which is not the main focus for this guidance) is important for "dockerizing" legacy applications that were originally developed with the traditional .NET Framework, like ASP.NET WebForms, MVC web apps or WCF (W ...
```
(which is not the main focus for this guidance) -> is important to day this? Perhaps say on  a sentence after this one?

CESARDELATORRE replied:
OK, removed.

mairaw commented on docs/standard/microservices-architecture/net-core-net-framework-containers/net-core-container-scenarios.md:
```diff
@@ -23,28 +23,26 @@ Clearly, if your goal is to have an application (web application or service) tha
+
+[Visual Studio for Mac](https://www.visualstudio.com/vs/visual-studio-mac/) is an IDE, evolution of Xamarin Studio, running in macOS and supports Docker since mid-2017.
+
+You can also use [Visual Studio Code](https://code.visualstudio.com/) (VS Code) on macOS, Linux, and Windows. VS Code fully supports .NET Core, including IntelliSense and debugging. Because VS Code is a lightweight editor, you can use it to develop containerized apps on the Mac in conjunction with the Docker CLI and the .NET Core CLI (dotnet cli). You can also target .NET Core with most third-par ...
```
Is it Sublime or Sublime Text?

.NET Core CLI or .NET Core command-line tools? Referring to the same thing with different names confuse people

Is it only OmniSharp that provides IntelliSense?

CESARDELATORRE replied:
OK, changed to the official name: .NET Core command-line interface (CLI) tools:
https://docs.microsoft.com/en-us/dotnet/core/tools/?tabs=netcore2x

mairaw commented on docs/standard/microservices-architecture/net-core-net-framework-containers/net-core-container-scenarios.md:
```diff
@@ -23,28 +23,26 @@ Clearly, if your goal is to have an application (web application or service) tha
 ## Creating and deploying microservices on containers
 
-You could use the full .NET framework for microservices-based applications (without containers) when using plain processes, because .NET Framework is already installed and shared across processes. However, if you are using containers, the image for .NET Framework (Windows Server Core plus the full .NET Framework within each image) is probably too heavy for a microservices-on-containers approach.
+Even when you could use the full .NET framework for microservices-based applications (without containers) when using plain processes, because .NET Framework is already installed and shared across processes. However, if you are using containers, the image for .NET Framework (Windows Server Core plus the full .NET Framework within each image) is probably too heavy for a microservices-on-containers  ...
```
we don't refer to framework as full

CESARDELATORRE replied:
OK, fixed.

mairaw reviewed (changes requested):
Looks good @CESARDELATORRE. Left a few comments. 

@BillWagner do we have an issue open to address the SEO issues we previously discussed for the e-books? E.g. "missing" descriptions, image filenames, images alt text, etc.

yishengjin1413 committed:
:mag_right: fix missing enum field (#3371)

Commit:
Removed quotation marks from titles, corrected typo (#3450)

* Update index.md

* Update introduction.md

* Update index.md

* feedback

bruceburge committed:
Update Form1.cs (#3444)

Removed unneeded iconForFile = Icon.ExtractAssociatedIcon(file.FullName); this would cause extraction every time and defeat the purpose of the the imagelist check.

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+---
```
OK, changed to lowercase, the filename itself plus the references to it.

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+keywords: Docker, Microservices, ASP.NET, Container, API Gateway
+author: CESARDELATORRE
+ms.author: wiwagn
+ms.date: 05/26/2017
```
OK, date updated.

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+
+**Figure 4-12**. Using a direct client-to-microservice communication architecture
+
+In this approach. each microservice has a public endpoint, sometimes with a different TCP port for each microservice. An example of an URL for a particular service could be the following URL in Azure:
```
OK, fixed.

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+
+In a production environment based on a cluster, that URL would map to the load balancer used in the cluster, which in turn distributes the requests across the microservices. In production environments, you could have an Application Delivery Controller (ADC) like [Azure Application Gateway](https://docs.microsoft.com/azure/application-gateway/application-gateway-introduction) between your microser ...
+
+A direct client-to-microservice communication architecture could be good enough for a small microservice-based applications, especially if the client app is a server-side web application like an ASP.NET MVC app. However, when you build large and complex microservice-based applications (for example, when handling dozens of microservice types), and expecially when the client apps are remote mobile  ...
```
Fixed.

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+
+Sometimes a granular API Gateway can also be a microservice by itself, and even have a domain or business name and related data. Having the API Gateway’s boundaries dictated by the business or domain will help you to get a better design.
+
+Granularity in the API Gateway tier can be especially useful for more advanced composite UI applications based on microservices, because the concept of a fine-grained API Gateway is similar to an UI composition service. We discuss this later in the [Creating composite UI based on microservices](#creating-composite-ui-based-on-microservices-including-visual-ui-shape-and-layout-generated-by-multipl ...
```
fixed

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+
+**Figure 4-14**. Using Azure API Management for your API Gateway
+
+In this case, when using a product like Azure API Management, the fact that you might have a single API Gateway is not so risky because this kind of API Gateways are "thinner", meaning that you don't implement custom C# code that could evolve towards a monolithic component. On the contrary, this type of product acts more like a reverse proxy where you can also filter the APIs from the internal mi ...
```
typo fixed

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+
+With Azure API Management, you can secure your APIs using a key, a token, and IP filtering. These features let you enforce flexible and fine-grained quotas and rate limits, modify the shape and behavior of your APIs using policies, and improve performance with response caching.
+
+In this guide and the reference sample application (eShopOnContainers) we are limiting the architecture to a simpler and custom-made containerized architecture in order to focus on plain containers without using PaaS products like Azure API Management. But for large microservice-based applications that are deployed into Microsoft Azure, we encourage you to review and adopt Azure API Management as ...
```
done.

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+
+## Drawbacks of the API Gateway pattern
+
+-   The most important drawback is that when you implement an API Gateway, you are coupling that tier with the internal microservices. Coupling like this might introduce serious difficulties for your application. (The cloud architect Clemens Vaster refers to this potential difficulty as “the new ESB” in his "[Messaging and Microservices](https://www.youtube.com/watch?v=rXi5CLjIQ9k)" session from  ...
```
OK, fixed.

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/direct-client-to-microservice-communication-versus-the-API-Gateway-pattern.md:
```diff
@@ -0,0 +1,117 @@
+
+-   An API Gateway can introduce increased response time due to the additional network call. However, this extra call usually has less impact than having a client interface that is too chatty directly calling the internal microservices.
+
+-   The API Gateway can represent a possible bottleneck if it is not scaled out properly
```
OK, fixed.

CESARDELATORRE committed:
Updates based on Maira's feedback.

CESARDELATORRE committed:
Updates based on Maira's feedback.

Commit:
Update Random2.cs (#3458)

mairaw committed:
broken link + xref format (#3457)

mairaw committed:
removed FSharp.Core info from APIs (#3449)

mairaw committed:
add new 4.7.1 API content (#3453)

* add new 4.7.1 API content

* feedback

mairaw committed:
fixed formatting (#3460)

Commit:
Added Windows Server, version 1709 (#3461)

guardrex committed:
Apply casing to artifacts.xml example (#2938)

@mairaw per https://github.com/dotnet/sdk/pull/1487

BillWagner commented on docs/standard/microservices-architecture/net-core-net-framework-containers/general-guidance.md:
```diff
@@ -25,15 +25,23 @@ In short, when you create new containerized .NET applications, you should consid
 -   You need to use third-party .NET libraries or NuGet packages that are not available for .NET Core.
 
-Using .NET Framework on Docker can improve your deployment experiences by minimizing deployment issues. This "lift and shift" scenario is important for "dockerizing" legacy applications (at least, those that are not based on microservices).
+Using .NET Framework on Docker can improve your deployment experiences by minimizing deployment issues. This [*"lift and shift" scenario*](https://aka.ms/liftandshiftwithcontainersebook) is important for "dockerizing" legacy applications that were originally developed with the traditional .NET Framework, like ASP.NET WebForms, MVC web apps or WCF (Windows Communication Foundation) services.
```
nit: I usually see "containerizing" preferred to "dockerizing"

CESARDELATORRE replied:
👍  💯

BillWagner reviewed (changes requested):
Overall, really good updates. I found a few small typos and issues that should be addressed. (A  few are nits and questions, which you can take or leave as you decide).

BillWagner commented:
@mairaw Issue #1083 is the issue for SEO.

CESARDELATORRE committed:
Update in Design and Implementation of Microservices

CESARDELATORRE committed:
Update in Design and Implementation of Microservices

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/scalable-available-multi-container-microservice-applications.md:
```diff
@@ -4,40 +4,50 @@ description: .NET Microservices Architecture for Containerized .NET Applications
 It looks like a logical approach. But how are you handling load-balancing, routing, and orchestrating these composed applications?
 
-The Docker CLI meets the needs of managing one container on one host, but it falls short when it comes to managing multiple containers deployed on multiple hosts for more complex distributed applications. In most cases, you need a management platform that will automatically start containers, suspend them or shut them down when needed, and ideally also control how they access resources like the ne ...
+The plain Docker Engine in single Docker hosts meets the needs of managing single image instances on one host, but it falls short when it comes to managing multiple containers deployed on multiple hosts for more complex distributed applications. In most cases, you need a management platform that will automatically start containers, sacale-out containers with multiple instances per image, suspend  ...
```
👍

CESARDELATORRE commented on docs/standard/microservices-architecture/architect-microservice-container-applications/scalable-available-multi-container-microservice-applications.md.saved.bak:
```diff
@@ -0,0 +1,137 @@
+---
```
👍

CESARDELATORRE commented on docs/standard/microservices-architecture/docker-application-development-process/docker-app-development-workflow.md:
```diff
@@ -43,7 +43,7 @@ However, just because Visual Studio makes those steps automatic does not mean th
 ## Step 1. Start coding and create your initial application or service baseline
 
-Developing a Docker application is similar to the way you develop an application without Docker. The difference is that while developing for Docker, you are deploying and testing your application or services running within Docker containers in your local environment (either a Linux VM or a Windows VM).
+Developing a Docker application is similar to the way you develop an application without Docker. The difference is that while developing for Docker, you are deploying and testing your application or services running within Docker containers in your local environment (either a Linux VM setup by Docker or directly Windows if using Windows Containers).
```
👍

CESARDELATORRE commented on docs/standard/microservices-architecture/docker-application-development-process/docker-app-development-workflow.md:
```diff
@@ -93,14 +93,14 @@ This action on a project (like an ASP.NET Web application or Web API service) ad
 You usually build a custom image for your container on top of a base image you can get from an official repository at the [Docker Hub](https://hub.docker.com/) registry. That is precisely what happens under the covers when you enable Docker support in Visual Studio. Your Dockerfile will use an existing aspnetcore image.
 
-Earlier we explained which Docker images and repos you can use, depending on the framework and OS you have chosen. For instance, if you want to use ASP.NET Core and Linux, the image to use is microsoft/aspnetcore:1.1. Therefore, you just need to specify what base Docker image you will use for your container. You do that by adding FROM microsoft/aspnetcore:1.1 to your Dockerfile. This will be auto ...
+Earlier we explained which Docker images and repos you can use, depending on the framework and OS you have chosen. For instance, if you want to use ASP.NET Core (Linux or Windows), the image to use is microsoft/aspnetcore:2.0. Therefore, you just need to specify what base Docker image you will use for your container. You do that by adding FROM microsoft/aspnetcore:2.0 to your Dockerfile. This wil ...
```
Usually it is better to have control over the specific version you are working with. For instance, when creating a project in VS 2017 with a .NET Core 2.0 project, when enabling Docker it will use the :2.0 tag, not the :latest. Specific versions are more predictable. :latest could. eventually, break your app or not.

CESARDELATORRE commented on docs/standard/microservices-architecture/docker-application-development-process/docker-app-development-workflow.md:
````diff
@@ -113,7 +113,7 @@ COPY ${source:-obj/Docker/publish} .
 ```
 
-In this case, the container is based on version 1.1 of the official ASP.NET Core Docker image for Linux; this is the setting FROM microsoft/aspnetcore:1.1. (For further details about this base image, see the [ASP.NET Core Docker Image](https://hub.docker.com/r/microsoft/aspnetcore/) page and the [.NET Core Docker Image](https://hub.docker.com/r/microsoft/dotnet/) page.) In the Dockerfile, you als ...
+In this case, the container is based on version 2.0 of the official ASP.NET Core Docker image (multi-arch for Linux and Windows). This is the setting FROM microsoft/aspnetcore:2.0. (For further details about this base image, see the [ASP.NET Core Docker Image](https://hub.docker.com/r/microsoft/aspnetcore/) page and the [.NET Core Docker Image](https://hub.docker.com/r/microsoft/dotnet/) page.) I ...
````
👍

CESARDELATORRE commented on docs/standard/microservices-architecture/docker-application-development-process/docker-app-development-workflow.md:
```diff
@@ -125,24 +125,36 @@ You can specify additional configuration settings in the Dockerfile, depending o
 
--   microsoft/dotnet:1.1-runtime
-    .NET Core 1.1 runtime-only on Linux Debian
+If you specify a tag, targeting a platform, that is explicit, like in the following cases:
```
👍 💯 (Adding "100" to track when I also updated the original WORD docs file for the .PDF generation)...

CESARDELATORRE committed:
Updates based on Bill's feedback

CESARDELATORRE committed:
Updates based on Bill's feedback

garg000dhruv committed:
Update language reference for non-trailing named arguments (#3307)

* Change example code explaining named arguments

Since we're expanding named arguments with C# 7.2, we need to have a
more sophisticated example to cover all the variants that are now
allowed along with ones that are still illegal. So instead of having a
2-parameter method, we'll have a 3-parameter method.

[Issue #: 3240]

* Change parameter name isShipped to hasShipped

* Update documentation for named arguments

Leveraging the updated code snippet, the named arguments section has
been updated with examples to explain non-trailing named arguments.

* Change code sample for named arguments

Based on PR feedback, the method in the code sample has been changed
to take in parameters that are the same type (string) to further amplify
the need of named arguments in the documentation.

* Update documentation based on PR feedback

* Point out C# 7.2 example with an explicit note

Per the review feedback, a note has been added when mentioning the C#
7.2 specific addition to named arguments so that it is explicitly clear
to the reader what the example entails.

* Specify the compiler error thrown for the bad example

* Add ArgumentException sample to code snippet

Per PR feedback, an ArgumentException was added to the example method
since both sellerName and productName are string types. This further
illustrates the point that named arguments enhance the readability of
the code.

* Add named arguments to exception in code snippet

BillWagner commented:
@mairaw This is ready from my perspective. Do you want to check and make sure you're ready to approve it as well?

carolinacmoravia committed:
CC6494: Update interpolated-strings.md #41 [AUTOMATED] (#3456)

* CC6494: Update interpolated-strings.md #41 [AUTOMATED]   

Hi, @BillWagner, 
This proposed file change comes from: https://github.com/dotnet/docs.ko-kr/pull/41/files. 
Could you help to review this and merge this?
Many thanks in advance.

* Update interpolated-strings.md

* Update interpolated-strings.md

* feedback

* Update interpolated-strings.md

carolinacmoravia committed:
M5294: Possible incorrect link name (#3470)

Hello, @dotnet-bot @dotnetcontent,
It seems that reported link should read "Windows Driver Kit" according to the title of the web site. Could you review this and merge if agreed, please? 
Many thanks in advance.

SimonCropp committed:
Update preprocessor-if.md (#3468)

kendrahavens committed:
Add Live Unit Testing link (#3465)

* Add Live Unit Testing link

Live Unit Testing link updated to point to the [Visual Studio doc on this topic](https://docs.microsoft.com/en-us/visualstudio/test/live-unit-testing-start?tabs=csharp) instead of being empty.

* Changed to relative link

pr-yemibedu committed:
Substitute positive enforcement (#3469)

Hello,
 Substitute positive enforcement: "In this regard" for "However". The containing sentence enforces the notion in the sentence prior to it. "However" would have indicated an exception to the rule. "In this regard" carries forward the realized example to the notion being covered. Thank you. Good day.

alexandair committed:
Update net-container-os-targets.md (#3459)

Fix typos. Remove "Windows". It's just "Nano Server". Image should be updated as well.

mairaw committed:
fixed some broken xref links (#3466)

* removed weird chars

* fixed broken link

CESARDELATORRE commented:
@mairaw Is everything ready to merge or is anything missing from my side?
Thank you! 👍

mairaw committed:
fixing second incorrect link name (#3471)

Commit:
Fix invalid regular expression (#3467)

Regular expression used \w in the part the matches the domain. A word (\w) matches with '_' which is an invalid character in a domain name.

mairaw commented:
@CESARDELATORRE I'll check. Can you resolve the merge conflict?

CESARDELATORRE committed:
commit due to rebase

CESARDELATORRE committed:
Merge branch 'master' of https://github.com/dotnet-architecture/docs

CESARDELATORRE commented:
I think I resolved the merge conflict. Can you confirm?

mairaw commented:
@CESARDELATORRE it now brought a bunch of commits that don't belong to this PR which is undesirable. let me try to clean this up. the change was simple enough that the resolve conflicts button was enabled here.

CESARDELATORRE commented:
I first pressed on the button but then it ididn't show me any possible further action, so I tried with the Git CLI, but it got a bit unclear, at some point..

mairaw commented:
Closing in favor of #3486
